pkg/operator/util/ca_test.go

```diff
@@ -0,0 +1,186 @@
+
+	// Create the pre-rotation CA
+	oldCAConfig, err := crypto.MakeSelfSignedCAConfig("foo", SigningCertificateLifetimeInDays)
+	if err != nil {
+		t.Fatalf("Failed to generate a new cert")
+	}
+	oldCA := &crypto.CA{
+		SerialGenerator: &crypto.RandomSerialGenerator{},
+		Config:          oldCAConfig,
+	}
+	oldBundle := []*x509.Certificate{oldCAConfig.Certs[0]}
+
+	// Generate a service cert with the pre-rotation CA
+	oldServingCert, err := oldCA.MakeServerCert(
+		hosts,
+		crypto.DefaultCertificateLifetimeInDays,
+	)
+	if err != nil {
+		t.Fatalf("Error generating server old service cert: %v", err)
+	}
+
+	// Rotate the CA
+	newSigningCA, err := RotateSigningCA(oldCAConfig.Certs[0], oldCAConfig.Key.(*rsa.PrivateKey))
+	if err != nil {
+		t.Fatalf("Error rotating signing ca: %v", err)
+	}
+	newCA := &crypto.CA{
+		SerialGenerator: &crypto.RandomSerialGenerator{},
+		Config:          newSigningCA.Config,
+	}
+	newBundle := newSigningCA.Bundle
+
+	// Generate a service cert with the post-rotation CA
+	newServingCert, err := newCA.MakeServerCert(
+		hosts,
+		crypto.DefaultCertificateLifetimeInDays,
+	)
+	if err != nil {
+		t.Fatalf("Error generating new service cert: %v", err)
+	}
+	// Append the intermediate cert to ensure that clients with the
+	// old bundle will be able to validate the service cert signed by
+	// the new ca.
+	newServingCert.Certs = append(newServingCert.Certs, newSigningCA.IntermediateCA)
+
+	testCases := map[string]struct {
+		servingCert *crypto.TLSCertificateConfig
+		bundle      []*x509.Certificate
+	}{
+		"Pre-rotation": {
+			servingCert: oldServingCert,
+			bundle:      oldBundle,
+		},
+		"Server rotated": {
+			servingCert: newServingCert,
+			bundle:      oldBundle,
+		},
+		"Client refreshed": {
+			servingCert: oldServingCert,
+			bundle:      newBundle,
+		},
+		"Server rotated and client refreshed": {
+			servingCert: newServingCert,
+			bundle:      newBundle,
+		},
+	}
+	for testName, tc := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			checkClientTrust(t, tc.servingCert, tc.bundle)
+		})
+	}
+}
+
+func checkClientTrust(t *testing.T, servingCert *crypto.TLSCertificateConfig, bundleCerts []*x509.Certificate) {
+	// Configure a server with the serving cert
+	rawCAs := [][]byte{}
+	for _, caCert := range servingCert.Certs {
+		rawCAs = append(rawCAs, caCert.Raw)
+	}
+	srv := http.Server{
+		TLSConfig: &tls.Config{
+			Certificates: []tls.Certificate{
+				{
+					Certificate: rawCAs,
+					PrivateKey:  servingCert.Key,
+				},
+			},
+		},
+	}
+
+	// Create a listener on a random port
+	listenerAddress := "127.0.0.1:0"
+	ln, err := net.Listen("tcp", listenerAddress)
+	if err != nil {
+		t.Fatalf("net.Listen: %v", err)
+	}
+	serverAddress := ln.Addr().String()
+	defer ln.Close()
+
+	// Start a server configured with the cert
+	go func() {
+		if err := srv.ServeTLS(ln, "", ""); err != nil && err != http.ErrServerClosed {
+			t.Fatalf("ServeTLS failed: %v", err)
+		}
+	}()
```

Comment:
Go merely passes the `srv.Addr` to `net.Listen` so the outcome should be the same - https://golang.org/src/net/http/server.go?s=97796:97864#L3098

I agree that this is a good candidate for an e2e test, but the unit test should eventually have just the verification part (which is what happens on that https server anyway, but I get your point) :+1:

Reply:
The test starts a server on a port guaranteed not to be already bound to ensure reliable execution in a test environment, which is achieved by using an address of `127.0.0.1:0`. As you say, it should be possible to set `srv.Addr` to that address and have the server start on a random port. However, `srv.Addr` is not updated once `ListenAndServeTLS` starts the server - the bound port is only set on the listener internal to that call chain. Hence the reason I'm starting the listener in advance - to enable reliable binding to a random port _and_ retrieval of that port to construct the address the client should target.

In my previous comment I said `I think it would be fine to switch to simple verification before merge`, and by that I meant replacing the https-based testing with cert verification. 

If we wanted to retain https-based testing, I'm not sure why it would be desirable to do that in e2e since verifying the cert chain of trust doesn't depend on a cluster. I'm a firm believer in running tests in the lowest-cost environment possible, and unit testing is about as cheap as we get (this test as currently written has sub-second execution time).

Reply:
I've updated the unit test to perform cert validation, and rewritten the e2e test to validate the trust with http server+client. The e2e test is not yet passing, I'm working on debugging the operator locally.